### PR TITLE
First step to support FreeBSD, OpenBSD, NetBSD, DragonFly BSD and Solaris

### PR DESCRIPTION
**Pull request by ghost**
Tested at OpenBSD. libref_gl.so fails to load.

### CODE REVIEW

**Comment by ghost** [comment]
@viciious can you please backport these into warsow_2.1 branch?

File: source/unix/unix_net.c [hunk 1]
```diff
@@ -81,9 +81,19 @@ int64_t Sys_NET_SendFile( socket_handle_t handle, int fileno, size_t offset, siz
 	len = count;
 	ssize_t result = sendfile( fileno, handle, _offset, &len, NULL, 0 );
 	result = len;
-#else
+#elif defined ( __linux__ )
 	ssize_t result = sendfile( handle, fileno, &_offset, count );
 	len = result;
+#else
```
Comment (viciious) [inline source/unix/unix_net.c:89]: At least FreeeBSD implements proper sendfile function

Comment (ghost) [reply source/unix/unix_net.c:89]: @viciious DragonFly BSD implements it too, but I have no idea how to use it.

Comment (jbeich) [reply source/unix/unix_net.c:89]: `sendfile()` on [FreeBSD](https://man.freebsd.org/sendfile/2) looks similar to [macOS](https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/sendfile.2.html) except `off_t len` (4th arg) is split into `off_t nbytes` (4th arg) + `off_t *sbytes` (6th arg). Looking at [libevent](https://github.com/libevent/libevent/blob/1f4f8769c4e78b14e6aa21d8300f432f67def8df/buffer.c#L2501-L2512) conditionals it'd be something like

```c
#elif defined ( __DragonFly__ ) || defined ( __FreeBSD__ )
	ssize_t result = sendfile( fileno, handle, _offset, count, NULL, &len, 0 );
```

Comment (ghost) [reply source/unix/unix_net.c:89]: @jbeich thank you!

**Comment by ghost** [comment]
@viciious to complete *BSD support, I need help to fix runtime. How to debug libref_gl.so to figure out why it fails to load? How to make qfusion look for data files, for example, in /usr/local/share/warsow by default without having to cd in that directory?

File: source/CMakeLists.txt [hunk 2]
```diff
@@ -215,7 +213,7 @@ if (NOT GAME_MODULES_ONLY)
         find_package(OpenAL)
 
         # We need SDL only on non-Windows platforms or when with SDL2-backend
-        if (USE_SDL2 OR (${CMAKE_SYSTEM_NAME} MATCHES "Linux"))
+        if (USE_SDL2 OR (${CMAKE_SYSTEM_NAME} MATCHES "Linux|BSD|DragonFly"))
```
Comment (jbeich) [inline source/CMakeLists.txt]: Maybe replace `(${CMAKE_SYSTEM_NAME} MATCHES "Linux|BSD|DragonFly")` with `UNIX AND NOT APPLE` to help Solaris.

File: source/unix/unix_net.c [hunk 3]
```diff
@@ -81,9 +81,22 @@ int64_t Sys_NET_SendFile( socket_handle_t handle, int fileno, size_t offset, siz
 	len = count;
 	ssize_t result = sendfile( fileno, handle, _offset, &len, NULL, 0 );
 	result = len;
-#else
+#elif defined ( __linux__ )
 	ssize_t result = sendfile( handle, fileno, &_offset, count );
 	len = result;
+#elif defined ( __DragonFly__ ) || defined ( __FreeBSD__ )
+	ssize_t result = sendfile( fileno, handle, _offset, count, NULL, &len, 0 );
+	result = len;
```
Comment (jbeich) [inline source/unix/unix_net.c]: `result = len` isn't necessary. If `sendfile` fails then `len` value is undefined. See outside the conditional line 101 to 104.

**Review by jbeich — COMMENTED** [review]
FreeBSD doesn't like `-D_POSIX_C_SOURCE=200112L -D_XOPEN_SOURCE=500`:
```c
FAILED: tv_server/CMakeFiles/qfusiontv_server.dir/__/qcommon/net.c.o
/usr/bin/cc -DDEDICATED_ONLY -DTV_MODULE_HARD_LINKED -DTV_SERVER_ONLY -DUSE_SDL2 -D_LARGEFILE64_SOURCE -D_LARGEFILE_SOURCE -I/usr/local/include -pipe -Wall -Wno-unused-function -fvisibility=hidden -Winvalid-pch -std=gnu99 -D_POSIX_C_SOURCE=200112L -D_XOPEN_SOURCE=500 -O3 -DNDEBUG -fno-strict-aliasing -ffast-math -funroll-loops -msse2 -O2 -MD -MT tv_server/CMakeFiles/qfusiontv_server.dir/__/qcommon/net.c.o -MF tv_server/CMakeFiles/qfusiontv_server.dir/__/qcommon/net.c.o.d -o tv_server/CMakeFiles/qfusiontv_server.dir/__/qcommon/net.c.o   -c qcommon/net.c
qcommon/net.c:104:6: warning: implicit declaration of function 'gethostname' is invalid in C99 [-Wimplicit-function-declaration]
        if( gethostname( hostname, 256 ) == SOCKET_ERROR ) {
            ^
qcommon/net.c:165:31: error: invalid application of 'sizeof' to an incomplete type 'struct sockaddr_in6'
                        memset( sadr_in6, 0, sizeof( *sadr_in6 ) );
                                                   ^~~~~~~~~~~~~
qcommon/net.c:163:11: note: forward declaration of 'struct sockaddr_in6'
                        struct sockaddr_in6 *sadr_in6 = (struct sockaddr_in6 *)sadr;
                               ^
qcommon/net.c:166:12: error: incomplete definition of type 'struct sockaddr_in6'
                        sadr_in6->sin6_family = AF_INET6;
                        ~~~~~~~~^
qcommon/net.c:163:11: note: forward declaration of 'struct sockaddr_in6'
                        struct sockaddr_in6 *sadr_in6 = (struct sockaddr_in6 *)sadr;
                               ^
qcommon/net.c:167:12: error: incomplete definition of type 'struct sockaddr_in6'
                        sadr_in6->sin6_port = na6->port;
                        ~~~~~~~~^
qcommon/net.c:163:11: note: forward declaration of 'struct sockaddr_in6'
                        struct sockaddr_in6 *sadr_in6 = (struct sockaddr_in6 *)sadr;
                               ^
qcommon/net.c:168:12: error: incomplete definition of type 'struct sockaddr_in6'
                        sadr_in6->sin6_scope_id = na6->scope_id;
                        ~~~~~~~~^
qcommon/net.c:163:11: note: forward declaration of 'struct sockaddr_in6'
                        struct sockaddr_in6 *sadr_in6 = (struct sockaddr_in6 *)sadr;
                               ^
qcommon/net.c:169:21: error: incomplete definition of type 'struct sockaddr_in6'
                        memcpy( &sadr_in6->sin6_addr, na6->ip, sizeof( sadr_in6->sin6_addr ) );
                                 ~~~~~~~~^
qcommon/net.c:163:11: note: forward declaration of 'struct sockaddr_in6'
                        struct sockaddr_in6 *sadr_in6 = (struct sockaddr_in6 *)sadr;
                               ^
qcommon/net.c:169:59: error: incomplete definition of type 'struct sockaddr_in6'
                        memcpy( &sadr_in6->sin6_addr, na6->ip, sizeof( sadr_in6->sin6_addr ) );
                                                                       ~~~~~~~~^
qcommon/net.c:163:11: note: forward declaration of 'struct sockaddr_in6'
                        struct sockaddr_in6 *sadr_in6 = (struct sockaddr_in6 *)sadr;
                               ^
qcommon/net.c:204:30: error: incomplete definition of type 'struct sockaddr_in6'
                        memcpy( na6->ip, &sadr_in6->sin6_addr, sizeof( na6->ip ) );
                                          ~~~~~~~~^
qcommon/net.c:200:17: note: forward declaration of 'struct sockaddr_in6'
                        const struct sockaddr_in6 *sadr_in6 = (const struct sockaddr_in6 *)s;
                                     ^
qcommon/net.c:205:24: error: incomplete definition of type 'struct sockaddr_in6'
                        na6->port = sadr_in6->sin6_port;
                                    ~~~~~~~~^
qcommon/net.c:200:17: note: forward declaration of 'struct sockaddr_in6'
                        const struct sockaddr_in6 *sadr_in6 = (const struct sockaddr_in6 *)s;
                                     ^
qcommon/net.c:206:28: error: incomplete definition of type 'struct sockaddr_in6'
                        na6->scope_id = sadr_in6->sin6_scope_id;
                                        ~~~~~~~~^
qcommon/net.c:200:17: note: forward declaration of 'struct sockaddr_in6'
                        const struct sockaddr_in6 *sadr_in6 = (const struct sockaddr_in6 *)s;
                                     ^
qcommon/net.c:227:50: error: invalid application of 'sizeof' to an incomplete type 'struct sockaddr_in6'
        addrlen = ( sockaddress.ss_family == AF_INET6 ? sizeof( struct sockaddr_in6 ) : sizeof( struct sockaddr_in ) );
                                                        ^     ~~~~~~~~~~~~~~~~~~~~~~~
qcommon/net.c:227:65: note: forward declaration of 'struct sockaddr_in6'
        addrlen = ( sockaddress.ss_family == AF_INET6 ? sizeof( struct sockaddr_in6 ) : sizeof( struct sockaddr_in ) );
                                                                       ^
qcommon/net.c:243:26: error: use of undeclared identifier 'PF_INET6'
        int protocol = ( ipv6 ? PF_INET6 : PF_INET );
                                ^
qcommon/net.c:243:37: error: use of undeclared identifier 'PF_INET'
        int protocol = ( ipv6 ? PF_INET6 : PF_INET );
                                           ^
qcommon/net.c:380:43: error: invalid application of 'sizeof' to an incomplete type 'struct sockaddr_in6'
        addrlen = ( addr.ss_family == AF_INET6 ? sizeof( struct sockaddr_in6 ) : sizeof( struct sockaddr_in ) );
                                                 ^     ~~~~~~~~~~~~~~~~~~~~~~~
qcommon/net.c:380:58: note: forward declaration of 'struct sockaddr_in6'
        addrlen = ( addr.ss_family == AF_INET6 ? sizeof( struct sockaddr_in6 ) : sizeof( struct sockaddr_in ) );
                                                                ^
qcommon/net.c:593:42: error: use of undeclared identifier 'SO_NOSIGPIPE'
        setsockopt( socket->handle, SOL_SOCKET, SO_NOSIGPIPE, &opt_val, sizeof( opt_val ) );
                                                ^
qcommon/net.c:601:42: error: use of undeclared identifier 'SO_NOSIGPIPE'
        setsockopt( socket->handle, SOL_SOCKET, SO_NOSIGPIPE, &opt_val, sizeof( opt_val ) );
                                                ^
qcommon/net.c:643:50: error: invalid application of 'sizeof' to an incomplete type 'struct sockaddr_in6'
        addrlen = ( sockaddress.ss_family == AF_INET6 ? sizeof( struct sockaddr_in6 ) : sizeof( struct sockaddr_in ) );
                                                        ^     ~~~~~~~~~~~~~~~~~~~~~~~
qcommon/net.c:643:65: note: forward declaration of 'struct sockaddr_in6'
        addrlen = ( sockaddress.ss_family == AF_INET6 ? sizeof( struct sockaddr_in6 ) : sizeof( struct sockaddr_in ) );
                                                                       ^
qcommon/net.c:1395:48: error: use of undeclared identifier 'in6addr_loopback'
                        return ( memcmp( address->address.ipv6.ip, &in6addr_loopback.s6_addr, sizeof( address->address.ipv6.ip ) ) == 0 ) ? true : false;
                                                                    ^
qcommon/net.c:1411:48: error: use of undeclared identifier 'in6addr_any'
                        return ( memcmp( address->address.ipv6.ip, &in6addr_any.s6_addr, sizeof( address->address.ipv6.ip ) ) == 0 ) ? true : false;
                                                                    ^
1 warning and 18 errors generated.
```
bumping to `-D_POSIX_C_SOURCE=200809L -D_XOPEN_SOURCE=600` doesn't help as some stuff isn't in POSIX:
```
qcommon/net.c:243:26: error: use of undeclared identifier 'PF_INET6'
        int protocol = ( ipv6 ? PF_INET6 : PF_INET );
                                ^
qcommon/net.c:243:37: error: use of undeclared identifier 'PF_INET'
        int protocol = ( ipv6 ? PF_INET6 : PF_INET );
                                           ^
qcommon/net.c:593:42: error: use of undeclared identifier 'SO_NOSIGPIPE'
        setsockopt( socket->handle, SOL_SOCKET, SO_NOSIGPIPE, &opt_val, sizeof( opt_val ) );
                                                ^
qcommon/net.c:601:42: error: use of undeclared identifier 'SO_NOSIGPIPE'
        setsockopt( socket->handle, SOL_SOCKET, SO_NOSIGPIPE, &opt_val, sizeof( opt_val ) );
                                                ^
4 errors generated.
```

File: source/unix/unix_net.c (hunk 3 above)
Comment (ghost) [reply source/unix/unix_net.c]: That is confusing, because that condition makes no sense, len is always equals result according to __APPLE__ and __linux__

**Comment by ghost** [comment]
Argh, I forgot about compiler arguments >.<

**Comment by jbeich** [comment]
> libref_gl.so fails to load.

Do you mean the following? It's caused by discrepancy between `CMAKE_SHARED_LIBRARY_SUFFIX` and gameshared/q_arch.h as used by `Com_LoadLibrary` or `FS_FOpenFile`. And the suffix is supposed to be unique enough, so the binaries can be put into pakfiles for all supported platforms.
```
Loading refresh module ref_gl... Not found libs/libref_gl_x86_64.so.
```

**Comment by ghost** [comment]
@jbeich wow, thank you for that.

**Comment by ghost** [comment]
It segfaulted, I can finally debug it, yeah!

**Comment by ghost** [comment]
Hm, if I play better with CMake, I'll be able to move BUILDSTRING, OSNAME, ARCH and CPUSTRING from q_arch.h to CMakeLists.txt

**Comment by ghost** [comment]
Huh, it appears the segfault was caused by me forgetting to rename libgame.so and libcgame.so

**Comment by ghost** [comment]
How about this?

**Comment by ghost** [comment]
Maybe Android will like this one...

**Comment by ghost** [comment]
Damn it, why it fails to define ARCH at Android?

**Comment by ghost** [comment]
@jbeich do we really have to use unique suffixes in libraries and executables if we are going to add Warsow and future qfusion games to ports? I mean it's ok if these binaries won't come for *BSD from official upstreams, because our ports systems are reliable, right?

File: source/ref_gl/r_cin.c [hunk 4]
```diff
@@ -19,6 +19,8 @@ Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */
 
 // r_cin.c
+#include <stdlib.h>
```
Comment (viciious) [inline source/ref_gl/r_cin.c]: stdlib.h is already included via q_arch.h via q_shared.h in gameshared, why is this needed?

File: source/ref_gl/r_cmds.c [hunk 5]
```diff
@@ -19,6 +19,8 @@ Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */
 
 // r_cmds.c
+#include <stdlib.h>
```
Comment (viciious) [inline source/ref_gl/r_cmds.c]: same thing here

File: source/ref_gl/r_cin.c (hunk 4 above)
Comment (ghost) [reply source/ref_gl/r_cin.c]: @viciious because for some reason I got warning that alloca() is not declared.

**Review by viciious — COMMENTED** [review]
That needs to be investigated. The include statement is clearly there in gameshared/q_arch.h

**Comment by ghost** [comment]
How the removal of 3 "#include <stdlib.h>" I added made qfusion fail to build at macOS?

**Comment by ghost** [comment]
Oh wait, that's a stalled build!

**Comment by ghost** [comment]
@viciious never mind that, somehow I managed to conditionally declare it in OpenBSD's <stdlib.h>

**Comment by ghost** [comment]
@viciious are you ok with patches which will just name all libraries and executables at OSes other than Windows, macOS, Linux and Android without suffixes like _freebsd_x86_64? I am asking this because I am a porter and I rely on ports systems, not official binaries. I expect qfusion games will be available at FreeBSD ports, OpenBSD ports, pkgsrc and Ravenports.

File: source/CMakeLists.txt [hunk 6]
```diff
@@ -289,8 +287,21 @@ if (${CMAKE_SYSTEM_NAME} MATCHES "Linux") # Do not add 'OR MINGW' here - we need
...
+    set(CMAKE_EXECUTABLE_SUFFIX ".${CMAKE_SYSTEM_PROCESSOR}${CMAKE_EXECUTABLE_SUFFIX}")
+
+elseif (UNIX AND NOT APPLE)
+    string(TOLOWER ${CMAKE_SYSTEM_NAME} OSNAME)
+    add_definitions(-DBUILDSTRING="${CMAKE_SYSTEM_NAME}")
+    add_definitions(-DOSNAME="${CMAKE_SYSTEM_NAME}")
+    add_definitions(-DARCH="${OSNAME}_${CMAKE_SYSTEM_PROCESSOR}")
+    add_definitions(-DCPUSTRING="${CMAKE_SYSTEM_PROCESSOR}")
+    set(CMAKE_SHARED_LIBRARY_SUFFIX "_${OSNAME}_${CMAKE_SYSTEM_PROCESSOR}${CMAKE_SHARED_LIBRARY_SUFFIX}")
+    set(CMAKE_EXECUTABLE_SUFFIX ".${OSNAME}_${CMAKE_SYSTEM_PROCESSOR}${CMAKE_EXECUTABLE_SUFFIX}")
```
Comment (ghost) [inline source/CMakeLists.txt]: Any ideas why this causes link time error? Here's build log: https://github.com/mazocomp/openbsd-wip/blob/warsow/games/warsow/BUILD

Comment (viciious) [reply source/CMakeLists.txt]: Nope, sorry

**Comment by viciious** [comment]
> @viciious are you ok with patches which will just name all libraries and executables at OSes other than Windows, macOS, Linux and Android without suffixes like _freebsd_x86_64? I am asking this because I am a porter and I rely on ports systems, not official binaries. I expect qfusion games will be available at FreeBSD ports, OpenBSD ports, pkgsrc and Ravenports.

How would suffixes hurt in case of ports?

**Comment by ghost** [comment]
@viciious in ports you as a user don't want to call "warsow.openbsd_i386", you want to call "warsow" (and I have no intention to write a shell script) and as a porter you have to keep a list of installed files, so in my case I would write "bin/warsow.openbsd_${ARCH}" to make processor architecture abstract and in Ravenports I would go even further "bin/warsow.${OS}_${ARCH}" to also abstract operating system (currently Ravenports supports Linux, FreeBSD, DragonFly BSD, Solaris/illumos and Darwin/macOS). Also as you've already seen by build log I gave above CMake starts giving full library names to -l flag and I don't know how to fix that besides removing those lines.

File: source/CMakeLists.txt (hunk 6 above)
Comment (jbeich) [reply source/CMakeLists.txt]: CMakeCache.txt or build directory may hold some clues. One of pkg-config files, *-config scripts or CMake modules probably returns `libfreetype` and CMake blindly appends `-l`.

Does it affect only Warsow's fork of Qfusion?

Comment (ghost) [reply source/CMakeLists.txt]: @jbeich I was just compiling Warsow 2.1, not a fork of qfusion, and that link time error appears only when I add those lines (from 297 to 305). I'll try to build current qfusion with my patches now.

Comment (ghost) [reply source/CMakeLists.txt]: Argh, impossible, git refuses to check out submodules.

**Comment by ghost** [comment]
@viciious btw, I know another way to fix that bug: if I knew Warsow structure better, I would easily write portable, readable, simple and fast BSD Makefile which any Unix-like OS can use (I don't know is it available for Windows and Android or not)

**Comment by viciious** [comment]
> 
> 
> @viciious btw, I know another way to fix that bug: if I knew Warsow structure better, I would easily write portable, readable, simple and fast BSD Makefile which any Unix-like OS can use (I don't know is it available for Windows and Android or not)

Sorry, I kinda lost track here. What bug?

As far as the Makefile is concerned, I'm not really interested in maintaniing both CMake and BSD Makefiles. Not worth the investment.

**Comment by ghost** [comment]
@viciious link-time error caused by standard CMake's functions and variables:
```
elseif (UNIX AND NOT APPLE)
    string(TOLOWER ${CMAKE_SYSTEM_NAME} OSNAME)
    add_definitions(-DBUILDSTRING="${CMAKE_SYSTEM_NAME}")
    add_definitions(-DOSNAME="${CMAKE_SYSTEM_NAME}")
    add_definitions(-DARCH="${OSNAME}_${CMAKE_SYSTEM_PROCESSOR}")
    add_definitions(-DCPUSTRING="${CMAKE_SYSTEM_PROCESSOR}")
    set(CMAKE_SHARED_LIBRARY_SUFFIX "_${OSNAME}_${CMAKE_SYSTEM_PROCESSOR}${CMAKE_SHARED_LIBRARY_SUFFIX}")
    set(CMAKE_EXECUTABLE_SUFFIX ".${OSNAME}_${CMAKE_SYSTEM_PROCESSOR}${CMAKE_EXECUTABLE_SUFFIX}")
endif()
```
Somehow these lines at OpenBSD cause CMake to pass full library names to -l flag:
```
[319/709] : && /usr/local/pobj/warsow-2.1/bin/cc -fPIC -O2 -pipe -Wall -Wno-unused-function -fvisibility=hidden -Winvalid-pch -I/usr/X11R6/include -std=gnu99 -DNDEBUG -fno-strict-aliasing -ffast-math -funroll-loops -msse2  -Wl,--as-needed  -s -shared -o /usr/local/pobj/warsow-2.1/warsow_21_sdk/source/source/build/libs/libftlib_openbsd_amd64.so ftlib/CMakeFiles/ftlib.dir/ftlib.c.o ftlib/CMakeFiles/ftlib.dir/ftlib_draw.c.o ftlib/CMakeFiles/ftlib.dir/ftlib_main.c.o ftlib/CMakeFiles/ftlib.dir/ftlib_syscalls.c.o ftlib/CMakeFiles/ftlib.dir/__/gameshared/q_math.c.o ftlib/CMakeFiles/ftlib.dir/__/gameshared/q_shared.c.o -L/usr/X11R6/lib -llibfreetype.so.29.0 -Wl,-rpath-link,/usr/X11R6/lib:/usr/local/lib && :
FAILED: /usr/local/pobj/warsow-2.1/warsow_21_sdk/source/source/build/libs/libftlib_openbsd_amd64.so 
: && /usr/local/pobj/warsow-2.1/bin/cc -fPIC -O2 -pipe -Wall -Wno-unused-function -fvisibility=hidden -Winvalid-pch -I/usr/X11R6/include -std=gnu99 -DNDEBUG -fno-strict-aliasing -ffast-math -funroll-loops -msse2  -Wl,--as-needed  -s -shared -o /usr/local/pobj/warsow-2.1/warsow_21_sdk/source/source/build/libs/libftlib_openbsd_amd64.so ftlib/CMakeFiles/ftlib.dir/ftlib.c.o ftlib/CMakeFiles/ftlib.dir/ftlib_draw.c.o ftlib/CMakeFiles/ftlib.dir/ftlib_main.c.o ftlib/CMakeFiles/ftlib.dir/ftlib_syscalls.c.o ftlib/CMakeFiles/ftlib.dir/__/gameshared/q_math.c.o ftlib/CMakeFiles/ftlib.dir/__/gameshared/q_shared.c.o -L/usr/X11R6/lib -llibfreetype.so.29.0 -Wl,-rpath-link,/usr/X11R6/lib:/usr/local/lib && :
ld: error: unable to find library -llibfreetype.so.29.0
cc: error: linker command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.
*** Error 1 in . (/usr/ports/devel/cmake/cmake.port.mk:39 'do-build': @cd /usr/local/pobj/warsow-2.1/build-amd64 && exec /usr/bin/env -i VER...)
*** Error 2 in . (/usr/ports/infrastructure/mk/bsd.port.mk:2796 '/usr/local/pobj/warsow-2.1/build-amd64/.build_done')
*** Error 2 in /usr/ports/mystuff/games/warsow (/usr/ports/infrastructure/mk/bsd.port.mk:2466 'all')
```

**Comment by jbeich** [comment]
Qfusion builds fine on FreeBSD (using this PR). Without build artifacts (e.g., CMakeCache.txt) it's hard to guess what went wrong. Maybe /usr/local/share/cmake/Modules/FindFreetype.cmake stumbles on the way OpenBSD versions shared libraries.

**Comment by ghost** [comment]
@jbeich that doesn't happen only to freetype when I add those lines, that happens to almost all other system-wide libraries. Still if I remove those lines that bug disappears. You didn't see other libraries because the build order of ninja is undefined, so you've seen a random output and I am not in mood to compile over and over to just show other libraries.

**Comment by ghost** [comment]
@viciious fortunately I have CMake guru @Lyberta, she gave me idea why the current CMakeLists.txt files are wrong in this project, I am going to fix them.

**Comment by ghost** [comment]
```set(CMAKE_SHARED_LIBRARY_SUFFIX "_${OSNAME}_${CMAKE_SYSTEM_PROCESSOR}${CMAKE_SHARED_LIBRARY_SUFFIX}")``` line should be blamed, this is not even the proper use of CMake.

**Comment by ghost** [comment]
Argh, SDL_SetWindowIcon() causes segmentation fault.

**Comment by ghost** [comment]
@jbeich @viciious I finally did it! Warsow runs at OpenBSD! Yeah!

**Comment by viciious** [comment]
> SDL_SetWindowIcon

This sounds like a SDL bug to me

**Comment by ghost** [comment]
@viciious I'll disable it for OpenBSD until it gets a fix. Also I am done porting Warsow to OpenBSD: https://marc.info/?l=openbsd-ports&m=155783186513074&w=2

**Comment by ghost** [comment]
@viciious btw, why your chosen design of qfusion is to place all binaries and libraries in one tarball? I am asking because it's a terrible practice for both users and developers. It would be better for all to just provide binaries and libraries for every OS and CPU architecture in different tarballs without using these complicated suffixes in binaries and libraries like _netbsd_sparc _freebsd_i386 _dragonfly_amd64 which only create inconveniences to users and developers. Also I am sure only Windows needs official binaries, because all other systems have reliable package systems. We, porters and users of Unix-like systems, only want from you two things: source code and data files. Porters will create ports, then those ports will be packaged by package system's upstream and users will just install those binary packages, because that's the best practice and is most convenient.

**Comment by viciious** [comment]
> @viciious btw, why your chosen design of qfusion is to place all binaries and libraries in one tarball? I am asking because it's a terrible practice for both users and developers. It would be better for all to just provide binaries and libraries for every OS and CPU architecture in different tarballs without using these complicated suffixes in binaries and libraries like _netbsd_sparc _freebsd_i386 _dragonfly_amd64 which only create inconveniences to users and developers. Also I am sure only Windows needs official binaries, because all other systems have reliable package systems. We, porters and users of Unix-like systems, only want from you two things: source code and data files. Porters will create ports, then those ports will be packaged by package system's upstream and users will just install those binary packages, because that's the best practice and is most convenient.

That's because binaries found in official repositories and ports usually lag behind the official releases by several years. What works for command line utilities, office programs and whatnot doesn't work for games with rapid development cycles.

**Comment by ghost** [comment]
@viciious you were right about the bug in SDL2, I see a buffer overflow: https://github.com/mazocomp/openbsd-wip/blob/warsow/games/warsow/BACKTRACE

That problem with outdated binaries is usually at Linux distros which like stability. In other cases if such games are not painful to port, they usually will end up updated immediately. If there are games which are painful to port, they will remain outdated because it takes an effort to port them, Warsow is an example of such kind of game (especially after you deprecated FreeBSD support).

How about we reduce the pain by providing official binaries in different tarballs for each architecture and each OS and removing these ARCH suffixes from libraries and binaries and appending ARCH suffixes to tarballs? Right now you are misusing CMake (CMAKE_SHARED_LIBRARY_SUFFIX and CMAKE_EXECUTABLE_SUFFIX should not be modified, that creates undefined behavior and I already suffered it, I thought the problem was in incorrect target_link_libraries() usage, but I was wrong) and making a complexity in source files.

**Comment by viciious** [comment]
What you're suggesting simply isn't feasible and/or unrealistic:
- packaging and hosting game distributions for each possible architecture and OS takes a lot of time, effort and disk space
- no one is interested in doing that anyway
- there are no game titles in the works or planned

I'm OK with dropping architecture suffixes though simply because packaging game libraries in pk3 files was a stupid idea in the first place and because x64 windows/linux is all that matters nowadays. The rest can somehow sort itself out via port maintainers or whatever means there might be.

**Comment by ghost** [comment]
Oh, right, Android doesn't run CMake.

**Comment by ghost** [comment]
@viciious I was afraid you were going to mention disk space usage, but anyway you are right that to most people only 64 bit Windows and Linux matter. Also: why this Android in CI doesn't run CMake?

**Comment by viciious** [comment]
> @viciious I was afraid you were going to mention disk space usage, but anyway you are right that to most people only 64 bit Windows and Linux matter. Also: why this Android in CI doesn't run CMake?

Android has its own set of Makefiles for reasons of keeping the build system more compact and straightforward. The system is wildly different from the mainline ones.

**Pull request merged**